apis/v1beta1/gateway_types.go

```diff
@@ -0,0 +1,806 @@
+
+	// Status defines the current state of Gateway.
+	//
+	// +kubebuilder:default={conditions: {{type: "Scheduled", status: "Unknown", reason:"NotReconciled", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status GatewayStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// GatewayList contains a list of Gateways.
+type GatewayList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Gateway `json:"items"`
+}
+
+// GatewaySpec defines the desired state of Gateway.
+//
+// Not all possible combinations of options specified in the Spec are
+// valid. Some invalid configurations can be caught synchronously via a
+// webhook, but there are many cases that will require asynchronous
+// signaling via the GatewayStatus block.
+type GatewaySpec struct {
+	// GatewayClassName used for this Gateway. This is the name of a
+	// GatewayClass resource.
+	GatewayClassName ObjectName `json:"gatewayClassName"`
+
+	// Listeners associated with this Gateway. Listeners define
+	// logical endpoints that are bound on this Gateway's addresses.
+	// At least one Listener MUST be specified.
+	//
+	// Each listener in a Gateway must have a unique combination of Hostname,
+	// Port, and Protocol.
+	//
+	// An implementation MAY group Listeners by Port and then collapse each
+	// group of Listeners into a single Listener if the implementation
+	// determines that the Listeners in the group are "compatible". An
+	// implementation MAY also group together and collapse compatible
+	// Listeners belonging to different Gateways.
+	//
+	// For example, an implementation might consider Listeners to be
+	// compatible with each other if all of the following conditions are
+	// met:
+	//
+	// 1. Either each Listener within the group specifies the "HTTP"
+	//    Protocol or each Listener within the group specifies either
+	//    the "HTTPS" or "TLS" Protocol.
+	//
+	// 2. Each Listener within the group specifies a Hostname that is unique
+	//    within the group.
+	//
+	// 3. As a special case, one Listener within a group may omit Hostname,
+	//    in which case this Listener matches when no other Listener
+	//    matches.
+	//
+	// If the implementation does collapse compatible Listeners, the
+	// hostname provided in the incoming client request MUST be
+	// matched to a Listener to find the correct set of Routes.
+	// The incoming hostname MUST be matched using the Hostname
+	// field for each Listener in order of most to least specific.
+	// That is, exact matches must be processed before wildcard
+	// matches.
+	//
+	// If this field specifies multiple Listeners that have the same
+	// Port value but are not compatible, the implementation must raise
+	// a "Conflicted" condition in the Listener status.
+	//
+	// Support: Core
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=64
```

Comment:
So this is odd to me. Why a max of 64 listeners? That seems arbitrary?

Reply:
We haven't written this down anywhere, I don't think, but we prefer not to have unbounded lists, since people will inevitably try to stuff 1000 things in there.

For times when we don't have a good idea, we do just pick an arbitrary limit that seems larger than what we think most uses will need, and accept that we may need to increase later. This is the case for all the limits you've called out here. Yes, they are arbitrary, but we needed a value >1 and < infinity, and we have no data about actual usage, so any number we pick is arbitrary.

Reply:
I just realized that I didn't make clear that the other reason for the limits is the max record size in etcd, which defaults to about 1MB of text in deserialized form.